src/components/images/draw.js

```diff
@@ -84,31 +84,37 @@ module.exports = function draw(gd) {
             var img = new Image();
             this.img = img;
 
-            // If not set, a `tainted canvas` error is thrown
-            img.setAttribute('crossOrigin', 'anonymous');
-            img.onerror = errorHandler;
-            img.onload = function() {
-                var canvas = document.createElement('canvas');
-                canvas.width = this.width;
-                canvas.height = this.height;
+            if(d.source && d.source.slice(0, 5) === 'data:') {
+                img.src = d.source;
```

Comment:
I see that you're setting `img.src` here even though the `Image` instance isn't used in this case. That's not a bad idea as this makes the early return a few lines above

```js
        if(this.img && this.img.src === d.source) {
            return;
        }
```

work.

But setting `img.src` spins off the `img.onload`, `img.onerror`, etc event chain. I'm not sure how if that can lead to performance losses, but it would be easy to stash the `d.source` data in something like `img._src` instead to skip potential `img.onload` / `img.onerror` work.

Reply:
We could even go one step further by

- if `source` is a data uri, do not wrap logic in a `Promise` (and do not append `gd._promises`)
- if `source` is a data uri, do not instantiate `Image`
- use `thisImage.attr('xlink:href')` as the "stash" value instead of `img.src`

Reply:
Yeah, that's a good point. `img._src` sounds good.